discovery_engine_core/core/src/utils.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//! Collection of utility functions.
+
+use std::{borrow::Cow, error::Error, fmt::Display, path::Path};
+
+use tracing::error;
+
+/// Like [`tokio::fs::remove_file()`] but doesn't fail if the file doesn't exist.
+pub(crate) async fn remove_file_if_exists(
+    path: impl AsRef<Path> + Send,
+) -> Result<(), std::io::Error> {
+    match tokio::fs::remove_file(path).await {
+        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(()),
+        other => other,
+    }
+}
+
+pub(crate) trait MiscErrorExt<T, E> {
+    // Logs the contained error if there is one.
+    fn log_error(self) -> Self
+    where
+        E: Display;
+
+    /// Like [`std::result::Result::ok()`] but if `slot` is `None` the error is moved into `slot`.
+    fn extract_first_error(self, slot: &mut Option<E>) -> Option<T>;
+
+    /// Like [`std::result::Result::ok()`] the error is moved into the passed in vec (using `push`).
+    fn extract_error(self, push_to: &mut Vec<E>) -> Option<T>;
+}
+
+impl<T, E> MiscErrorExt<T, E> for Result<T, E> {
+    fn log_error(self) -> Self
+    where
+        E: Display,
+    {
+        if let Err(err) = &self {
+            error!(error = %err);
+        }
+        self
+    }
+
+    fn extract_first_error(self, slot: &mut Option<E>) -> Option<T> {
+        match self {
+            Ok(val) => Some(val),
+            Err(err) => {
+                if slot.is_none() {
+                    *slot = Some(err);
+                }
+                None
+            }
+        }
+    }
+
+    fn extract_error(self, push_to: &mut Vec<E>) -> Option<T> {
+        match self {
+            Ok(val) => Some(val),
+            Err(err) => {
+                push_to.push(err);
+                None
+            }
+        }
+    }
```

Comment:
I cycled through using `log_error`, `extract_first_error` and `extract_error` when writing this PR, in the end I ended up only using `extract_error` but I thing the other ones are helpful, too so I kept them. (e.g. `foo().log_error().ok()` for ignoring but logging errors is pretty neat).